### PR TITLE
add graphql websockets support (TT-904)

### PR DESCRIPTION
**Pull request by pvormste**
This PR adds support for websockets using the GraphQL websockets protocol.
To be able to use this feature, websockets need to be enabled for the gateway.

I have also set up a GraphiQL with websockets support that you can use for testing: 
https://github.com/pvormste/graphiql-websockets-tyk

Instructions are located in the readme.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

### CODE REVIEW

File: gateway/reverse_proxy.go [hunk 1]
```diff
@@ -714,59 +729,83 @@ func (rt *TykRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
...
-	if !p.TykAPISpec.GraphQL.Enabled {
-		res, err = roundTripper.RoundTrip(outreq)
-		return
+	if p.TykAPISpec.GraphQL.Enabled {
+		res, hijacked, err = p.handleGraphQL(roundTripper, outreq, w)
+		return res, hijacked, latency, err
 	}
 
-	// process graphql
+	res, hijacked, err = p.sendRequestToUpstream(roundTripper, outreq)
```
Comment (spetrunin) [inline gateway/reverse_proxy.go:743]: You could remove hijacked from `sendRequestToUpstream` method
```
res, err = p.sendRequestToUpstream(roundTripper, outreq)
return
```

File: gateway/reverse_proxy.go [hunk 2]
```diff
@@ -714,59 +729,83 @@ func (rt *TykRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
...
 
-	// execution mode
+	var isIntrospection bool
+	isIntrospection, err = gqlRequest.IsIntrospectionQuery()
+	if err != nil {
+		return
+	}
+
+	if isIntrospection {
+		return p.handleGraphQLIntrospection()
```
Comment (spetrunin) [inline gateway/reverse_proxy.go:766]: same could be done here
```
res, err = p.handleGraphQLIntrospection()
return
```

**Review by spetrunin — CHANGES_REQUESTED** [review]

File: gateway/mw_graphql.go [hunk 3]
```diff
@@ -129,6 +135,17 @@ func (m *GraphQLMiddleware) ProcessRequest(w http.ResponseWriter, r *http.Reques
 		return errors.New("there was a problem proxying the request"), http.StatusInternalServerError
 	}
 
+	if config.Global().HttpServerOptions.EnableWebSockets && websocket.IsWebSocketUpgrade(r) {
```
Comment (spetrunin) [inline gateway/mw_graphql.go:138]: what if simplify conditions?
```go
	if websocket.IsWebSocketUpgrade(r) {
		if !config.Global().HttpServerOptions.EnableWebSockets {
			return errors.New("websockets are not allowed"), http.StatusUnprocessableEntity
		}

		if !m.websocketUpgradeUsesGraphQLProtocol(r) {
			return errors.New("invalid websocket protocol for upgrading to a graphql websocket connection"), http.StatusBadRequest
		}

		ctxSetGraphQLIsWebSocketUpgrade(r, true)
		return nil, http.StatusSwitchingProtocols
	}
```

**Review by spetrunin — APPROVED** [review]

**Review by furkansenharputlu — APPROVED** [review]
Looks good!

**Comment by pvormste** [comment]
manually merged to release-3

**Pull request merged**